### PR TITLE
[ABI] Initial draft of canonicalization and minimization of generic signatures

### PR DESCRIPTION
**Pull request by DougGregor**
This is a rough, initial draft describing the canonicalization and minimization of generic signatures for the purposes of the ABI. Comments greatly appreciated!

Addresses the documentation part of [SR-3733](https://bugs.swift.org/browse/SR-3733) and rdar://problem/31412994.

### CODE REVIEW

File: docs/ABI/GenericSignature.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+## Canonicalization
+
+A generic signature is *canonical* when each of its constraints is [canonical](#canonical-constraints) and the entries in the generic signature appear in canonical order.
+
+1. Generic type parameters (that are not nested types) are listed first 
+   ordered by [type parameter ordering](#type-parameter-ordering)
+2. Constraints follow, ordered first by the 
+   [type parameter ordering](#type-parameter-ordering) of the left-hand  
+   operand and then by constraint kind. The left-hand side of a constraint
+   is always a type parameter `T`, which can be a generic parameter of a
```
Comment (slavapestov) [inline docs/ABI/GenericSignature.md]: "a generic parameter _or_ a"?

File: docs/ABI/GenericSignature.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+1. Generic type parameters (that are not nested types) are listed first 
+   ordered by [type parameter ordering](#type-parameter-ordering)
+2. Constraints follow, ordered first by the 
+   [type parameter ordering](#type-parameter-ordering) of the left-hand  
+   operand and then by constraint kind. The left-hand side of a constraint
+   is always a type parameter `T`, which can be a generic parameter of a 
+   nested type thereof (e.g., `T.SubSequence.Iterator.Element`). 
+   Constraints are ordered as follows:
+   1. A superclass constraint `T: C`, where `C` is a class.
+   2. A layout constraints (e.g., `T: some-layout`), where the right-hand
```
Comment (slavapestov) [inline docs/ABI/GenericSignature.md]: "A layout constraint"

File: docs/ABI/GenericSignature.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+A layout or conformance constraint is canonical when its left-hand side is the anchor of its equivalence class. A superclass constraint is canonical when its left-hand side is the anchor of its equivalence class and its right-hand side is a canonical concrete (class) type. Same-type constraint canonicalization is discussed in detail in the following section, but some basic rules apply: the left-hand side is always a type parameter, and the right-hand side is either a type parameter that follows the left-hand side (according to the [type parameter ordering](#type-parameter-ordering)) or is a canonical concrete type.
+
+### Same-type constaints
+
+The canonical form of superclass, layout, and conformance constraints are trivially canonicalized using the anchor of the appropriate equivalence class. Same-type constraints, on the other hand, are responsible for forming those equivalence classes. Let's expand our running example to include a third `Collection`:
```
Comment (slavapestov) [inline docs/ABI/GenericSignature.md]: What about superclass constraints where the right hand side is a generic class?

**Comment by slavapestov** [comment]
It's not clear to me how we canonicalize concrete structural types that contain type parameters inside of them, eg `T.Foo == (T.Bar, T.Baz)`. Does that fall out from the other rules or does it deserve a mention? Either way an example would be good.

File: docs/ABI/GenericSignature.md (hunk 1 above)
Comment (DougGregor) [reply docs/ABI/GenericSignature.md]: Got it, thanks.

File: docs/ABI/GenericSignature.md (hunk 2 above)
Comment (DougGregor) [reply docs/ABI/GenericSignature.md]: Got it, thanks.

File: docs/ABI/GenericSignature.md (hunk 3 above)
Comment (DougGregor) [reply docs/ABI/GenericSignature.md]: Yeah, I should call that out. The RHS needs to be a canonical concrete type.

**Comment by DougGregor** [comment]
Canonicalizing concrete structural types containing type parameters needs to be called out. I'll push an update shortly that does so, thanks!

**Comment by DougGregor** [comment]
I think this is in "good enough" shape to merge. Additional comments are certainly welcome!

**Pull request merged**